pkg/http/api_crypto.go

```diff
@@ -0,0 +1,194 @@
+/*
+Copyright 2023 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package http
+
+import (
+	"bytes"
+	"io"
+	"strings"
+
+	"github.com/valyala/fasthttp"
+
+	contribCrypto "github.com/dapr/components-contrib/crypto"
+	"github.com/dapr/dapr/pkg/messages"
+	"github.com/dapr/dapr/utils"
+	"github.com/dapr/kit/ptr"
+	encv1 "github.com/dapr/kit/schemes/enc/v1"
+)
+
+func (a *api) constructCryptoEndpoints() []Endpoint {
+	// These APIs are not implemented as Universal because the gRPC APIs are stream-based.
+	return []Endpoint{
+		{
+			Methods: []string{fasthttp.MethodPut},
+			Route:   "crypto/{name}/encrypt",
+			Version: apiVersionV1alpha1,
+			Handler: a.onCryptoEncrypt,
+		},
+		{
+			Methods: []string{fasthttp.MethodPut},
+			Route:   "crypto/{name}/decrypt",
+			Version: apiVersionV1alpha1,
+			Handler: a.onCryptoDecrypt,
+		},
+	}
+}
+
+// Handler for crypto/<component-name>/encrypt
+// Query-string args:
+// - keyName (required)
+// - algorithm (required)
+// - omitKeyName
+// - decryptionKeyName
+// - cipher
+func (a *api) onCryptoEncrypt(reqCtx *fasthttp.RequestCtx) {
+	// Get the component
+	componentName := reqCtx.UserValue(nameParam).(string)
+	component, err := a.cryptoGetComponent(componentName)
+	if err != nil {
+		// Error has been logged already
+		universalFastHTTPErrorResponder(reqCtx, err)
+		return
+	}
+
+	// Get the required properties from the query string
+	keyName := string(reqCtx.QueryArgs().Peek("keyName"))
+	if keyName == "" {
+		err = messages.ErrBadRequest.WithFormat("missing query string parameter 'keyName'")
+		log.Debug(err)
+		universalFastHTTPErrorResponder(reqCtx, err)
+		return
+	}
+	algorithm := string(reqCtx.QueryArgs().Peek("algorithm"))
```

Comment:
are query Args the best place to have these values? 
Should it not be part of the POST request?

Reply:
This API doesn't have a POST body :( Because it's (eventually, when #6248 is merged) supposed to work with streams, it treats the body as the message to encrypt, entirely. The body is not encoded as JSON (it's can be a large file and forcing the use of JSON would require a lot of memory and additional processing time, and would make it impossible to work with the data as a stream), nor it uses multipart/form-data (because if the other fields, besides the file, are placed after the file, then we need to buffer the entire file in memory before we can start processing it).

Reply:
My concern is that query args are directly available to be cached and used. Can we have the key names and algorithms in some other place?

Reply:
Done